### PR TITLE
Added ignoreVersion and loaded plugin list to StatTracking for issue #1444

### PR DESCRIPTION
**Pull request by madman2021**
Cleaned up some duplicated code and added tracking for Loaded plugins and status of the "ignoreversion" flag for #1444

This will probably need some changes to the back-end at stats.tshock.co but I'm not sure on how that's implemented

### CODE REVIEW

**Comment by WhiteXZ** [comment]
@Celant Could you have a look at this too please?

**Comment by Celant** [comment]
@WhiteXZ Just had a look, I can't see any issues with it. Good clean implementation.

@madman2021 I assume this has been tested? If so, I'm happy to check the backend stuff tonight to see if it's coming through properly.

**Comment by madman2021** [comment]
@Celant I've tested yes. here is the output without URL Encoding and formatted for visual appeal.
```json
{
    "port": 7777,
    "currentPlayers": 0,
    "maxPlayers": 8,
    "systemRam": 24,
    "version": "4.3.23.0",
    "ignorePluginVersion": false,
    "loadedPlugins": [
        "CRITICAL HIT!:1.1",
        "TShock:4.3.23.0"
    ],
    "terrariaVersion": "v1.3.5.3",
    "providerId": "",
    "serverId": "f2397d20-4a2f-4553-9021-e3285ee22441",
    "mono": false
}
```

**Comment by Ijwu** [comment]
I'd like to request that you include reflection based scanning determining the API version of the plugins loaded into the startup of TShock.

I don't find there being a use in knowing if a server has used `-ignoreversion` so much as knowing _which_ plugins are loaded through it.

**Comment by madman2021** [comment]
@Ijwu point taken ill amend my commit

**Comment by Ijwu** [comment]
Thanks, @madman2021, super awesome of you to contribute this. Much appreciated.

**Comment by hakusaro** [comment]
I still think just knowing if they set the flag is a valid attribute

**Comment by Ijwu** [comment]
@hakusaro I don't disagree, I just don't think it provides as much use. If we know what the most popular, yet out of date, plugin is we can probably pre-empt issues with its usage.

**Comment by madman2021** [comment]
I've kept the flag in and replaced the Version with the Api version
I guess we don't really care about the assembly version just the API version

**Comment by Ijwu** [comment]
The plugin version is useful as well, due to the possibility of there being multiple plugin versions on the same api version.

**Comment by madman2021** [comment]
Changed as requested, added another commit

**Comment by Ijwu** [comment]
I would like to suggest that acceptance of this PR be contingent on the resolution of issue #1446. I hope this is acceptable. It is in no way an attempt to demean your contribution @madman2021. I appreciate your work.

**Comment by madman2021** [comment]
@Ijwu I'm cool do what you have to, I'm happy to contribute either way.
@Celant New Json output 
```json
{
    "port": 7777,
    "currentPlayers": 0,
    "maxPlayers": 8,
    "systemRam": 24,
    "version": "4.3.23.0",
    "ignorePluginVersion": false,
    "loadedPlugins": [
        {
            "name": "CRITICAL HIT!",
            "version": "1.1",
            "apiVersion": "2.1"
        },
        {
            "name": "EventStopper",
            "version": "1.0.1.1",
            "apiVersion": "2.1"
        },
        {
            "name": "TShock",
            "version": "4.3.23.0",
            "apiVersion": "2.1"
        }
    ],
    "terrariaVersion": "v1.3.5.3",
    "providerId": "",
    "serverId": "ef95b5c3-06a8-4b59-b06a-2e038adfbbba",
    "mono": false
}
```
@Celant I see the back-end code is in one of your public repository's, have you taken it upon yourself to update it if required to support the new structure or would you mind if I posted a PR to that repository I don't mind either way.
Thanks

**Comment by Celant** [comment]
@madman2021 Sorry, been quite busy. I've been rewriting the back-end of the stats server offline, so that repo is likely to change quite a lot fairly soon. You're welcome to submit a PR, and I'll re-work my offline one to also track the new data, or I can just do it offline and commit the offline stuff into the main repo :) (I'd ideally like to do the latter, I should have been committing often anyway!)

R.E. The JSON output, I prefer this new format a lot, thank you!

**Comment by madman2021** [comment]
@Celant No problem I'll leave you to it :)

**Comment by Celant** [comment]
@madman2021 Yep, I'm seeing the stats coming through to the back-end server with no issues. Extract from the logs is below.

https://gist.github.com/Celant/f95e97e89f5316f3fbcec9d6d829ad6c

**Comment by madman2021** [comment]
@Celant Nice, that's my live server. Forgot I left the debug version on there actually 💩

**Comment by Celant** [comment]
@madman2021 Sorry for this stagnating. Can you get this up to date with general-devel, and I'll make sure it gets merged. Are you in Discord? I've got some more changes that need making but they're best suited for a separate PR.

**Comment by Celant** [comment]
👍 Looks good to merge, validated the data on the backend.

File: TShockAPI/StatTracker.cs [hunk 1]
```diff
@@ -139,49 +140,81 @@ public void Start()
...
+				ignorePluginVersion = ServerApi.IgnoreVersion,
+				loadedPlugins = GetLoadedPlugins()
 			};
 		}
+
+		private PluginItem[] GetLoadedPlugins()
+		{
+			if (plugins != null)
+			{
+				return plugins;//Return early
```
Comment (hakusaro) [inline TShockAPI/StatTracker.cs]: Spaces go in between inline comments and semicolons.

File: TShockAPI/StatTracker.cs [hunk 2]
```diff
@@ -139,49 +140,81 @@ public void Start()
...
+			}
+
+			plugins = new PluginItem[ServerApi.Plugins.Count];//Initialize with enough room to store the ammount of plugins loaded.
```
Comment (hakusaro) [inline TShockAPI/StatTracker.cs]: Spaces go in between inline comments and semicolons.

File: TShockAPI/StatTracker.cs [hunk 3]
```diff
@@ -139,49 +140,81 @@ public void Start()
...
+				plugins[i] = pluginItem;
+			}
+			return plugins;
+		}
+
+		private long GetTotalSystemRam(bool isMono)
+		{
+			if (totalMem != 0)
+			{
+				return totalMem;//Return early
```
Comment (hakusaro) [inline TShockAPI/StatTracker.cs]: Spaces go in between inline comments and semicolons.

File: TShockAPI/StatTracker.cs [hunk 4]
```diff
@@ -139,49 +140,81 @@ public void Start()
...
+		}
+
+		private long GetTotalSystemRam(bool isMono)
+		{
+			if (totalMem != 0)
+			{
+				return totalMem;//Return early 
+			}
+
+			if (isMono)//Set totalMem so it can be returned later
```
Comment (hakusaro) [inline TShockAPI/StatTracker.cs]: Spaces go in between inline comments and semicolons.

**Pull request merged**